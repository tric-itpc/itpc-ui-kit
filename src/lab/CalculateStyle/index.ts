import type { CSSProperties, RefObject } from "react"

import {
  getDocumentDimensions,
  getElementDimensions,
  getParentDimensions,
} from "../getDemensions"

import { DEFAULT_DISTANCE_BETWEEN_ELEMENTS, DIVIDER_IN_TWO } from "./constants"
import {
  ALLOWED_POSITIONS,
  type GetHorizontalPositionArg,
  HORIZONTAL_POSITION,
  type PositionType,
  VERTICAL_POSITION,
} from "./types"

export const getHorizontalPosition = (
  arg: GetHorizontalPositionArg
): HORIZONTAL_POSITION => {
  const {
    defaultParentWidth,
    distanceRight,
    documentWidth,
    elementWidth,
    parentLeft,
    parentWidth,
    scrollbarWidth,
  } = arg

  const currentWidth = elementWidth !== 0 ? elementWidth : defaultParentWidth

  const rightEdgeFromCenter = parentWidth / DIVIDER_IN_TWO + distanceRight
  const leftEdgeFromCenter = parentWidth / DIVIDER_IN_TWO + parentLeft
  const calendarHalfWidth = elementWidth / DIVIDER_IN_TWO
  const rightStartParent = distanceRight + parentWidth
  const leftEndParent = parentWidth + parentLeft

  if (currentWidth !== undefined) {
    if (rightStartParent <= elementWidth && leftEndParent <= elementWidth) {
      return HORIZONTAL_POSITION.CALCULATED
    }

    if (parentWidth > currentWidth) {
      return HORIZONTAL_POSITION.LEFT
    } else {
      if (documentWidth - scrollbarWidth < currentWidth) {
        return HORIZONTAL_POSITION.CALCULATED
      } else {
        if (parentWidth + distanceRight > currentWidth) {
          return HORIZONTAL_POSITION.LEFT
        }

        if (
          rightEdgeFromCenter > calendarHalfWidth &&
          leftEdgeFromCenter > calendarHalfWidth
        ) {
          return HORIZONTAL_POSITION.CENTER
        }

        if (parentWidth + parentLeft > currentWidth) {
          return HORIZONTAL_POSITION.RIGHT
        }
        return HORIZONTAL_POSITION.CALCULATED
      }
    }
  }
  return HORIZONTAL_POSITION.CALCULATED
}

export const getVerticalPosition = (
  refParent: RefObject<HTMLDivElement>,
  ref: RefObject<HTMLDivElement | HTMLUListElement>,
  distanceBetweenElements?: number
): VERTICAL_POSITION => {
  const { documentHeight } = getDocumentDimensions()
  const parent = refParent?.current
  const { parentBottom } = getParentDimensions(parent as HTMLDivElement)
  const distanceUnder: number = documentHeight - parentBottom
  const currentDistanceBetweenElements =
    distanceBetweenElements ?? DEFAULT_DISTANCE_BETWEEN_ELEMENTS
  if (
    distanceUnder >
    (ref.current?.offsetHeight ?? 0) + currentDistanceBetweenElements
  ) {
    return VERTICAL_POSITION.BOTTOM
  } else {
    return VERTICAL_POSITION.TOP
  }
}

export const getCalculatePosition = (
  refParent: React.RefObject<HTMLDivElement>,
  refChildren: React.RefObject<HTMLDivElement | HTMLUListElement>,
  position: PositionType,
  distanceBetweenElements?: number,
  horizontalAlignment?: HORIZONTAL_POSITION
): CSSProperties => {
  if (
    position !== ALLOWED_POSITIONS.ABSOLUTE &&
    position !== ALLOWED_POSITIONS.FIXED
  ) {
    throw new Error(
      `Недопустимое значение свойства position: ${position}. Допустимые значения: absolute, fixed`
    )
  }

  if (refParent?.current && refChildren?.current) {
    const { documentHeight, documentWidth } = getDocumentDimensions()

    const { elementHeight, elementWidth } = getElementDimensions(
      refChildren.current
    )

    const { parentBottom, parentLeft, parentTop, parentWidth } =
      getParentDimensions(refParent.current)

    const scrollbarWidth: number = window.innerWidth - documentWidth

    const distanceRight: number = documentWidth - (parentLeft + parentWidth)

    const argHorizontalPosition: GetHorizontalPositionArg = {
      distanceRight,
      documentWidth,
      elementWidth,
      parentLeft,
      parentWidth,
      scrollbarWidth,
    }

    const horizontalPosition: HORIZONTAL_POSITION = horizontalAlignment
      ? horizontalAlignment
      : getHorizontalPosition(argHorizontalPosition)

    const distanceUnderInput: number = documentHeight - parentBottom

    const currentDistanceBetweenElements: number =
      distanceBetweenElements ?? DEFAULT_DISTANCE_BETWEEN_ELEMENTS

    let verticalPosition: VERTICAL_POSITION
    if (distanceUnderInput > elementHeight + currentDistanceBetweenElements) {
      verticalPosition = VERTICAL_POSITION.BOTTOM
    } else {
      verticalPosition = VERTICAL_POSITION.TOP
    }

    const scrollTop = window.scrollY || window.pageYOffset
    const verticalBottomPosition: number =
      parentBottom + currentDistanceBetweenElements

    const verticalBottomPositionAbsolute: number =
      parentBottom + currentDistanceBetweenElements + scrollTop

    const verticalTopPosition: number =
      parentTop - currentDistanceBetweenElements - elementHeight

    const verticalTopPositionAbsolute: number =
      parentTop - currentDistanceBetweenElements - elementHeight + scrollTop

    const horizontalCenterPosition: number =
      parentLeft - (elementWidth - parentWidth) / DIVIDER_IN_TWO

    const horizontalRightPosition: number =
      documentWidth - parentLeft - parentWidth

    const horizontalCalculatedPosition: number =
      documentWidth <= elementWidth
        ? 0
        : (documentWidth - elementWidth) / DIVIDER_IN_TWO

    const currentPosition = refChildren.current.style.position

    if (currentPosition.length && String(currentPosition) !== position) {
      refChildren.current.style.position = position
    }

    if (verticalPosition === VERTICAL_POSITION.BOTTOM) {
      const topPositionBottom =
        position === ALLOWED_POSITIONS.FIXED
          ? verticalBottomPosition
          : verticalBottomPositionAbsolute

      switch (horizontalPosition) {
        case HORIZONTAL_POSITION.LEFT:
          return {
            bottom: "auto",
            left: parentLeft,
            position,
            right: "auto",
            top: topPositionBottom,
          }
        case HORIZONTAL_POSITION.RIGHT:
          return {
            bottom: "auto",
            left: "auto",
            position,
            right: horizontalRightPosition,
            top: topPositionBottom,
          }
        case HORIZONTAL_POSITION.CENTER:
          return {
            bottom: "auto",
            left: horizontalCenterPosition,
            position,
            right: "auto",
            top: topPositionBottom,
          }
        case HORIZONTAL_POSITION.CALCULATED:
          return {
            bottom: "auto",
            left: horizontalCalculatedPosition,
            position,
            right: "auto",
            top: topPositionBottom,
          }
      }
    }

    if (verticalPosition === VERTICAL_POSITION.TOP) {
      const topPosition =
        position === ALLOWED_POSITIONS.FIXED
          ? verticalTopPosition
          : verticalTopPositionAbsolute

      switch (horizontalPosition) {
        case HORIZONTAL_POSITION.LEFT:
          return {
            bottom: "auto",
            left: parentLeft,
            position,
            right: "auto",
            top: topPosition,
          }
        case HORIZONTAL_POSITION.RIGHT:
          return {
            bottom: "auto",
            left: "auto",
            position,
            right: horizontalRightPosition,
            top: topPosition,
          }
        case HORIZONTAL_POSITION.CENTER:
          return {
            bottom: "auto",
            left: horizontalCenterPosition,
            position,
            right: "auto",
            top: topPosition,
          }
        case HORIZONTAL_POSITION.CALCULATED:
          return {
            bottom: "auto",
            left: horizontalCalculatedPosition,
            position,
            right: "auto",
            top: topPosition,
          }
      }
    }
  }

  return {}
}
