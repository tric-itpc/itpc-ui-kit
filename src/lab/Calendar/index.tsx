import React, { useEffect, useRef, useState } from "react"
import cn from 'classnames'

import { useOnClickOutside } from "../../_hooks"

import { Day } from "./types"
import {
  getCalendarDays,
  getTodayMonthYear,
  initCurrentDate,
  initCurrentTime,
  initDays
} from "./utils"
import {
  CalendarControl,
  CalendarTable,
  CalendarTimes,
  SelectionMenuMonth,
  SelectionMenuYear
} from "./components"

import './styles.css'

export interface Props {
  id: string
  currentValue?: string
  offsetYear?: number
  activeDates?: string[]
  disabledDates?: string[]
  disabledAfterDate?: string
  disabledBeforeDate?: string
  show: boolean
  handleShow: () => void
  onChange?: (
    date: string,
    event: React.ChangeEvent<HTMLInputElement> | React.MouseEvent<HTMLButtonElement> | React.MouseEvent<HTMLTableDataCellElement>
  ) => void
  withTime?: boolean
}

export const Calendar: React.FC<Props> = ({
  id,
  currentValue = getTodayMonthYear(),
  offsetYear = 10,
  activeDates,
  disabledDates,
  disabledAfterDate,
  disabledBeforeDate,
  show,
  handleShow,
  onChange,
  withTime = false
}: Props) => {
  const [currentDate, setCurrentDate] = useState<string>(initCurrentDate(currentValue, withTime))
  const [hours, setHours] = useState<string>(initCurrentTime(currentValue, withTime)[0])
  const [minutes, setMinutes] = useState<string>(initCurrentTime(currentValue, withTime)[1])
  const [seconds, setSeconds] = useState<string>(initCurrentTime(currentValue, withTime)[2])
  const [days, setDays] = useState<Day[]>(initDays(currentValue, withTime))
  const [isShowSelectMonth, setIsShowSelectMonth] = useState<boolean>(false)
  const [isShowSelectYear, setIsShowSelectYear] = useState<boolean>(false)

  const calendarRef = useRef<HTMLDivElement>(null)

  const changeCurrentDate = (date: string): void => {
    setCurrentDate(date)
    setDays(getCalendarDays(date))
    setIsShowSelectMonth(false)
    setIsShowSelectYear(false)
  }

  const changeCurrentTime = (time: string, event: React.ChangeEvent<HTMLInputElement> | React.MouseEvent<HTMLButtonElement>): void => {
    const [currentHours, currentMinutes, currentSeconds] = time.split(':')
    setHours(currentHours)
    setMinutes(currentMinutes)
    setSeconds(currentSeconds)

    if (onChange) {
      onChange(`${currentDate}T${time}`, event)
    }
  }

  const onChangeDate = (date: string, event: React.MouseEvent<HTMLTableDataCellElement>): void => {
    if (onChange) {
      if (!withTime) {
        onChange(date, event)
      }

      if (withTime) {
        onChange(`${date}T${hours}:${minutes}:${seconds}`, event)
      }
    }

    if (!withTime) {
      handleShow()
    }
  }

  const handleShowSelectMonth = (): void => {
    setIsShowSelectMonth(!isShowSelectMonth)
    setIsShowSelectYear(false)
  }

  const handleShowSelectYear = (): void => {
    setIsShowSelectYear(!isShowSelectYear)
    setIsShowSelectMonth(false)
  }

  useOnClickOutside(calendarRef, handleShow)

  useEffect(() => {
    if (currentValue !== `${currentDate}T${hours}:${minutes}:${seconds}`) {
      setCurrentDate(initCurrentDate(currentValue, withTime))
      setHours(initCurrentTime(currentValue, withTime)[0])
      setMinutes(initCurrentTime(currentValue, withTime)[1])
      setSeconds(initCurrentTime(currentValue, withTime)[2])
      setDays(initDays(currentValue, withTime))
    }
  }, [currentValue, withTime])

  return (
    <div
      className={cn(
        'itpc-calendar',
        show && 'itpc-calendar_opened'
      )}
      ref={calendarRef}
    >
      <CalendarControl
        currentDate={currentDate}
        changeCurrentDate={changeCurrentDate}
        handleShowSelectYear={handleShowSelectYear}
        handleShowSelectMonth={handleShowSelectMonth}
      />
      <CalendarTable
        id={id}
        days={days}
        currentDate={currentDate}
        activeDates={activeDates}
        disabledDates={disabledDates}
        disabledAfterDate={disabledAfterDate}
        disabledBeforeDate={disabledBeforeDate}
        onChange={onChangeDate}
      />
      {withTime && (
        <CalendarTimes
          hours={hours}
          minutes={minutes}
          seconds={seconds}
          onChange={changeCurrentTime}
        />
      )}
      {isShowSelectMonth && (
        <SelectionMenuMonth
          currentDate={currentDate}
          changeCurrentDate={changeCurrentDate}
        />
      )}
      {isShowSelectYear && (
        <SelectionMenuYear
          currentDate={currentDate}
          offsetYear={offsetYear}
          changeCurrentDate={changeCurrentDate}
        />
      )}
    </div>
  )
}
