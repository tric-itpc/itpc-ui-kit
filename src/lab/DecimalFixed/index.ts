import { Decimal } from "decimal.js-light"

import { IDecimalFixed } from "./types"

export class DecimalFixed implements IDecimalFixed {
  constructor(value: string, accuracy: null | number = null, empty = false) {
    const [int, fraction] = value.split(".")

    this.int = empty ? (int.length ? Number(int) : null) : Number(int)
    this.accuracy = accuracy === null || accuracy < 0 ? null : accuracy
    this.fraction = fraction ?? null

    this.empty = empty

    if (this.accuracy && this.fraction) {
      this.fraction =
        this.accuracy > 0 ? this.fraction.substring(0, this.accuracy) : null
    }
  }

  accuracy?: null | number
  empty: boolean
  fraction?: null | string
  int: null | number

  equals(value: Decimal): boolean {
    return new Decimal(this.generate()).equals(value)
  }

  generate(): string {
    return this.fraction === null
      ? this.empty && this.int === null
        ? ""
        : String(this.int)
      : `${this.int}.${this.fraction}`
  }

  greaterThan(value: Decimal): boolean {
    return new Decimal(this.generate()).greaterThan(value)
  }

  isEmpty(): boolean {
    return this.toString().length === 0
  }

  isValid(): boolean {
    if (Number.isNaN(this.int)) {
      return false
    }

    return !(this.fraction !== null && Number.isNaN(Number(this.fraction)))
  }

  isZero(): boolean {
    return new Decimal(this.generate()).isZero()
  }

  lessThan(value: Decimal): boolean {
    return new Decimal(this.generate()).lessThan(value)
  }

  minus(value: string): Decimal {
    return new Decimal(this.generate()).minus(value)
  }

  plus(value: string): Decimal {
    return new Decimal(this.generate()).plus(value)
  }

  toDecimal(): Decimal {
    return new Decimal(this.generate())
  }

  toString(): string {
    return this.generate()
  }
}
