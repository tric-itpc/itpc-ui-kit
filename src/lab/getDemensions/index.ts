import type {
  DocumentDimensions,
  ElementDimensions,
  ParentDimensions,
} from "../CalculateStyle/types"

export const getDocumentDimensions = (): DocumentDimensions => ({
  documentHeight: document.documentElement.clientHeight,
  documentWidth: document.documentElement.clientWidth,
})

export const getParentDimensions = (ref: HTMLDivElement): ParentDimensions => {
  const rect: DOMRect = ref.getBoundingClientRect()
  return {
    parentBottom: rect.bottom,
    parentHeight: rect.height,
    parentLeft: rect.left,
    parentTop: rect.top,
    parentWidth: rect.width,
  }
}

export const getElementDimensions = (
  ref: HTMLDivElement | HTMLUListElement
): ElementDimensions => ({
  elementHeight: ref.offsetHeight,
  elementWidth: ref.offsetWidth,
})
