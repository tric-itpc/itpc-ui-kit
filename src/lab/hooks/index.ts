export { useOnClickOutside } from "./useOnClickOutside"
export { useTheme } from "./useTheme"
export { useWindowSize } from "./useWindowSize"
