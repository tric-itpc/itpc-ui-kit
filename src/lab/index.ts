export { Calendar } from "./Calendar"
export { DecimalFixed } from "./DecimalFixed"
export type { IDecimalFixed } from "./DecimalFixed/types"
export * from "./hooks"
