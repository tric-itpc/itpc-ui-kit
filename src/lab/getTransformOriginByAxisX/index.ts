import type { RefObject } from "react"

import { getVerticalPosition } from "../CalculateStyle"
import { VERTICAL_POSITION } from "../CalculateStyle/types"

export const getTransformOriginByAxisX = (
  refParent: RefObject<HTMLDivElement>,
  refChildren: RefObject<HTMLDivElement | HTMLUListElement>,
  distanceBetweenElements?: number
): string => {
  const verticalPosition: VERTICAL_POSITION = getVerticalPosition(
    refParent,
    refChildren,
    distanceBetweenElements
  )

  const startPosition: VERTICAL_POSITION =
    verticalPosition === VERTICAL_POSITION.TOP
      ? VERTICAL_POSITION.BOTTOM
      : VERTICAL_POSITION.TOP

  return `center ${startPosition}`
}
