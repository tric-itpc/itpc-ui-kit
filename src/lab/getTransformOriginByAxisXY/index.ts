import { getHorizontalPosition } from "../CalculateStyle"
import { DIVIDER_IN_TWO } from "../CalculateStyle/constants"
import {
  type GetHorizontalPositionArg,
  HORIZONTAL_POSITION,
  VERTICAL_POSITION,
} from "../CalculateStyle/types"
import { DEFAULT_WIDTH_CALENDAR } from "../Calendar/constants"
import {
  getDocumentDimensions,
  getElementDimensions,
  getParentDimensions,
} from "../getDemensions"

export const getTransformOriginByAxisXY = (
  parentRef: React.RefObject<HTMLDivElement>,
  childrenRef: React.RefObject<HTMLDivElement>
): string => {
  if (parentRef?.current && childrenRef?.current) {
    const { parentBottom, parentLeft, parentWidth } = getParentDimensions(
      parentRef.current
    )
    const calculatedTransformPosition: number =
      parentLeft + parentWidth / DIVIDER_IN_TWO

    const { documentHeight, documentWidth } = getDocumentDimensions()
    const { elementHeight, elementWidth } = getElementDimensions(
      childrenRef.current
    )

    const distanceUnderParent: number = documentHeight - parentBottom

    let verticalPosition: VERTICAL_POSITION
    if (distanceUnderParent > elementHeight + 8) {
      verticalPosition = VERTICAL_POSITION.BOTTOM
    } else {
      verticalPosition = VERTICAL_POSITION.TOP
    }

    const distanceRight: number = documentWidth - (parentLeft + parentWidth)
    const scrollbarWidth: number = window.innerWidth - documentWidth

    const argHorizontalPosition: GetHorizontalPositionArg = {
      defaultParentWidth: DEFAULT_WIDTH_CALENDAR,
      distanceRight,
      documentWidth,
      elementWidth,
      parentLeft,
      parentWidth,
      scrollbarWidth,
    }

    const horizontalPosition: HORIZONTAL_POSITION = getHorizontalPosition(
      argHorizontalPosition
    )

    const startPosition: VERTICAL_POSITION =
      verticalPosition === VERTICAL_POSITION.TOP
        ? VERTICAL_POSITION.BOTTOM
        : VERTICAL_POSITION.TOP

    switch (horizontalPosition) {
      case HORIZONTAL_POSITION.LEFT:
        return `left ${startPosition}`
      case HORIZONTAL_POSITION.RIGHT:
        return `right ${startPosition}`
      case HORIZONTAL_POSITION.CENTER:
        return `center ${startPosition}`
      case HORIZONTAL_POSITION.CALCULATED:
        return `${calculatedTransformPosition}px ${startPosition}`

      default:
        return `center ${startPosition}`
    }
  } else {
    return `center`
  }
}
