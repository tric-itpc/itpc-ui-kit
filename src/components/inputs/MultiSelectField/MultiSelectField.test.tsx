import React from "react"

import { fireEvent, render, screen } from "@testing-library/react"

import { getSelectButton } from "../../../lab/tests"

import { MultiSelectField } from "./index"

describe("Компонент MultiSelectField", () => {
  const items = [
    { id: "1", value: "Элемент 1" },
    { id: "2", value: "Элемент 2" },
    { disabled: true, id: "3", value: "Элемент 3" },
  ]

  beforeAll(() => {
    window.HTMLElement.prototype.scrollIntoView = jest.fn()
  })

  test("Корректно рендерится", () => {
    const { container } = render(
      <MultiSelectField
        data-testid="multi-select-field"
        items={items}
        onChange={() => {}}
        placeholder="Выберите элементы"
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    expect(inputElement).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test("Открывает и закрывает выпадающий список", () => {
    render(
      <MultiSelectField
        data-testid="multi-select-field"
        items={items}
        onChange={() => {}}
        placeholder="Выберите элементы"
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)

    fireEvent.click(buttonElement)

    const firstItem = screen.getByText("Элемент 1")
    const parentElement = firstItem.closest(".itpc-list-box")

    expect(parentElement).toBeInTheDocument()
    expect(parentElement).toHaveClass("itpc-list-box_opened")
    expect(parentElement).not.toHaveClass("itpc-list-box_closed")

    fireEvent.click(buttonElement)

    expect(parentElement).toHaveClass("itpc-list-box_closed")
    expect(parentElement).not.toHaveClass("itpc-list-box_opened")
  })

  test("Выбирает и отменяет выбор элемента", () => {
    const handleChange = jest.fn()
    const { rerender } = render(
      <MultiSelectField
        items={[
          { id: "1", value: "Элемент 1" },
          { id: "2", value: "Элемент 2" },
        ]}
        data-testid="multi-select-field"
        onChange={handleChange}
        selectedItems={[]}
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)

    fireEvent.click(buttonElement)

    const firstItem = screen.getByText("Элемент 1")
    fireEvent.click(firstItem)

    expect(handleChange).toHaveBeenCalledTimes(1)
    expect(handleChange).toHaveBeenLastCalledWith(["1"])

    rerender(
      <MultiSelectField
        items={[
          { id: "1", value: "Элемент 1" },
          { id: "2", value: "Элемент 2" },
        ]}
        data-testid="multi-select-field"
        onChange={handleChange}
        selectedItems={["1"]}
      />
    )

    fireEvent.click(firstItem)

    expect(handleChange).toHaveBeenCalledTimes(2)
    expect(handleChange).toHaveBeenLastCalledWith([])
  })

  test("Выбирает несколько элементов", () => {
    const handleChange = jest.fn()
    const { rerender } = render(
      <MultiSelectField
        items={[
          { id: "1", value: "Элемент 1" },
          { id: "2", value: "Элемент 2" },
        ]}
        data-testid="multi-select-field"
        onChange={handleChange}
        selectedItems={[]}
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)

    fireEvent.click(buttonElement)

    const firstItem = screen.getByText("Элемент 1")
    fireEvent.click(firstItem)

    expect(handleChange).toHaveBeenCalledTimes(1)
    expect(handleChange).toHaveBeenLastCalledWith(["1"])

    rerender(
      <MultiSelectField
        items={[
          { id: "1", value: "Элемент 1" },
          { id: "2", value: "Элемент 2" },
        ]}
        data-testid="multi-select-field"
        onChange={handleChange}
        selectedItems={["1"]}
      />
    )

    const secondItem = screen.getByText("Элемент 2")
    fireEvent.click(secondItem)

    expect(handleChange).toHaveBeenCalledTimes(2)
    expect(handleChange).toHaveBeenLastCalledWith(["1", "2"])
  })

  test("Не позволяет выбрать отключенный элемент", () => {
    const handleChange = jest.fn()
    render(
      <MultiSelectField
        items={[
          { id: "1", value: "Элемент 1" },
          { disabled: true, id: "2", value: "Элемент 2" },
        ]}
        data-testid="multi-select-field"
        onChange={handleChange}
        selectedItems={[]}
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)

    fireEvent.click(buttonElement)

    const disabledItem = screen.getByText("Элемент 2")
    fireEvent.click(disabledItem)

    expect(handleChange).not.toHaveBeenCalled()
  })

  test("Обрабатывает навигацию с клавиатуры", () => {
    const handleChange = jest.fn()
    render(
      <MultiSelectField
        data-testid="multi-select-field"
        items={items}
        onChange={handleChange}
        placeholder="Выберите элементы"
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)
    fireEvent.click(buttonElement)

    fireEvent.keyDown(buttonElement, { key: "ArrowDown" })
    const firstItem = screen.getByText("Элемент 1")
    expect(firstItem).toHaveClass("itpc-select-item_active")

    fireEvent.keyDown(buttonElement, { key: "Enter" })
    expect(handleChange).toHaveBeenCalledWith(["1"])
  })

  test("Корректно отображает выбранные элементы", () => {
    render(
      <MultiSelectField
        data-testid="multi-select-field"
        items={items}
        onChange={() => {}}
        placeholder="Выберите элементы"
        selectedItems={["1", "2"]}
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)
    expect(buttonElement).toHaveTextContent("Выбрано 2 элементов")
  })

  test("Корректно работает в отключенном состоянии", () => {
    render(
      <MultiSelectField
        data-testid="multi-select-field"
        items={items}
        onChange={() => {}}
        placeholder="Выберите элементы"
        disabled
      />
    )

    const inputElement = screen.getByTestId("multi-select-field")
    const buttonElement = getSelectButton(inputElement)
    expect(buttonElement).toBeDisabled()
  })
})
