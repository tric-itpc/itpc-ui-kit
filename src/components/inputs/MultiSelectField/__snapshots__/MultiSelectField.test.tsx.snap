// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент MultiSelectField Корректно рендерится 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    class="itpc-multi-select  itpc-multi-select_hover"
    data-testid="multi-select-field"
    role="combobox"
    tabindex="0"
  >
    <button
      class="itpc-multi-select__button"
      type="button"
    >
      <label
        class="itpc-placeholder"
      >
        Выберите элементы
      </label>
    </button>
    <i
      class="itpc-icon__arrow itpc-icon__arrow_orientation_bottom"
    >
      <svg
        height="8"
        viewBox="0 0 12 8"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.5134 0L6 4.82292L1.48659 0L0 1.58854L6 8L12 1.58854L10.5134 0Z"
        />
      </svg>
    </i>
  </div>
</div>
`;
